config/cmake: find MPI if the MPI component is requested

mathstuf opened the pull request:
Otherwise projects using `conduit::conduit_mpi` may get errors about `MPI::MPI_C` not being a target without finding it themselves.

mathstuf committed:
config/cmake: find MPI if the MPI component is requested

Otherwise projects using `conduit::conduit_mpi` may get errors about
`MPI::MPI_C` not being a target without finding it themselves. Instead
of requesting users to find MPI themselves, let them request support via
an MPI component.

cyrush commented:
We made a choice that folks can use a conduit built with MPI, downstream w/o using its MPI features w/o having to locate mpi. 

We actually have tests that support this use case, one is our basic example of how to build conduit with cmake:

https://github.com/LLNL/conduit/blob/develop/src/examples/using-with-cmake/CMakeLists.txt

This means that downstream use must find MPI when using Conduit's MPI features.

The need to find mpi downstream is reflected in our using-with-cmake-mpi example:

https://github.com/LLNL/conduit/blob/dffc5a6c7b433665399231aa537469be32bce865/src/examples/using-with-cmake-mpi/CMakeLists.txt#L65

There are some other caveats about this: Even when MPI is on we may not search for it when it is implicitly supported by the compiler (cray use cases...)

This config var reveals if that is the case:

https://github.com/LLNL/conduit/blob/dffc5a6c7b433665399231aa537469be32bce865/src/config/ConduitConfig.cmake.in#L49

mathstuf commented:
Can it be requested that Conduit does it via a component?

`find_package(Conduit COMPONENTS MPI)`?

cyrush commented:
@mathstuf yes, that sounds like a much nicer way to formalize this

mathstuf committed:
config/cmake: also set `Conduit_FOUND`

The package is intended to be found via `find_package(Conduit)`, so set
the package name's `_FOUND` variable directly.

mathstuf committed:
config/cmake: handle components when finding Conduit

mathstuf committed:
examples/mpi: use the MPI component

Also drop the explicit `MPI::MPI_C` link as `conduit::conduit_mpi`
already does this.

mathstuf committed:
config/cmake: always search for Conduit

Two subsequent searches could have different component requests, so
remove the `CONDUIT_FOUND` guard that skips over that logic.

mathstuf commented:
OK, I've added logic to handle components and then added support for the MPI component which takes the considerations you mentioned into account. I also updated the example to use the MPI component and bail if it isn't reported as being found. The example was also unconditionally using `MPI::MPI_C` which isn't correct when MPI is implicit in the toolchain (i.e., the targets aren't being used), so that is also removed.

cyrush commented on src/config/ConduitConfig.cmake.in:
```diff
@@ -16,50 +16,75 @@ cmake_minimum_required(VERSION 3.8 FATAL_ERROR)
 
 @PACKAGE_INIT@
 
-if(NOT CONDUIT_FOUND)
```
Q: Was this bad practice before, or does new component based search work in a similar way?

mathstuf replied:
It was "meh" practice before, but benign because the logic was pretty much unconditional on what the user requested by this point (CMake should handle version constraints using the `ConduitVersion.cmake` file information).

cyrush replied:
Thanks for letting me know, and again thanks for this improvement to the approach.

cyrush reviewed (commented):
Awesome, thanks for adding the component based logic and updating the examples. 

I have one question about the cmake config changes.

mathstuf commented on src/config/ConduitConfig.cmake.in:
```diff
@@ -16,50 +16,75 @@ cmake_minimum_required(VERSION 3.8 FATAL_ERROR)
+set(Conduit_FOUND TRUE)
 
-    set(CONDUIT_FOUND TRUE)
+set(_conduit_known_components
```
Bah, this needs unset.

scottwittenburg commented:
Thanks folks, I'm looking forward to this landing! Is anything more needed here before merging?

cyrush commented:
@mathstuf  thanks for this!

@scottwittenburg merged today, sorry for the delay

scottwittenburg commented:
> merged today, sorry for the delay

Thanks and no worries!
